File: frontend/src/api/interface/vipanel.ts
export namespace ViPanel {
    export interface Caps {
        structuredEvents: boolean;
        resume: boolean;
        interrupt: boolean;
        auth: boolean;
        models: string[];
        effortLevels: string[];
        commands: { name: string; desc: string }[];
        // 登录方式由后端的 harness 声明。文案在 i18n 里按 id 取。
        // needsCodeInput 决定码往哪个方向走：true 是浏览器给码、粘回终端；
        // false 是终端给码、拿到别的设备上去输。
        loginModes: { id: string; needsCodeInput: boolean }[];
    }

    export interface Session {
        id: string;
        title: string;
        cwd: string;
        dir: string;
        harness: string;
        status: 'idle' | 'working' | 'unread' | 'sleeping' | 'error';
        alive: boolean;
        lastUsed: number;
        notes: string;
        mode: string;
        capabilities: Caps;
    }

    export interface Outbound {
        enabled: boolean;
        url: string;
    }

    export interface ReachResult {
        url: string;
        purpose: string;
        ok: boolean;
        status: number;
        detail: string;
    }

    export interface Account {
        id: string;
        label: string;
        // 和当前 live 配置一致的那个。由后端比对指纹得出，不是前端记的。
        active: boolean;
        addedAt: number;
    }

    export interface Prereq {
        binary: string;
        hint: string;
    }

    export interface InstallInfo {
        installable: boolean;
        // 本机缺少的前置依赖。非空时显示原因，不显示安装按钮。
        missing?: Prereq[];
        note?: string;
    }

    export interface Harness {
        id: string;
        displayName: string;
        // 这台机器上装没装。没装就别让用户点进去撞墙。
        installed: boolean;
        install: InstallInfo;
        capabilities: Caps;
    }

    export interface AuthState {
        supported: boolean;
        hookInstalled: boolean;
        loggedIn: boolean;
        authMethod?: string;
        email?: string;
        plan?: string;
    }

    export interface History {
        id: string;
        cwd: string;
        title: string;
        mtime: number;
        size: number;
    }

    export interface Pool {
        size: number;
        active: number;
        order: string[];
    }
}

File: agent/app/dto/vipanel.go
package dto

type ViSessionItem struct {
	ID           string   `json:"id"`
	Title        string   `json:"title"`
	Cwd          string   `json:"cwd"`
	Dir          string   `json:"dir"`
	Harness      string   `json:"harness"`
	Status       string   `json:"status"`
	Alive        bool     `json:"alive"`
	LastUsed     int64    `json:"lastUsed"`
	Notes        string   `json:"notes"`
	Mode         string   `json:"mode"`
	Capabilities ViCaps   `json:"capabilities"`
	_            struct{} `json:"-"`
}

type ViCaps struct {
	StructuredEvents bool     `json:"structuredEvents"`
	Resume           bool     `json:"resume"`
	Interrupt        bool     `json:"interrupt"`
	Auth             bool     `json:"auth"`
	Models           []string `json:"models"`
	EffortLevels     []string `json:"effortLevels"`
	Commands         any      `json:"commands"`
	LoginModes       any      `json:"loginModes"`
}

type ViHarnessItem struct {
	ID          string `json:"id"`
	DisplayName string `json:"displayName"`
	// Installed 为 false 时界面必须把它标成不可用，而不是让用户点进去撞墙。
	Installed bool          `json:"installed"`
	Install   ViInstallInfo `json:"install"`
	Caps      ViCaps        `json:"capabilities"`
}

// ViInstallInfo 告诉界面「这个 harness 能不能装、装之前还缺什么」。
type ViInstallInfo struct {
	Installable bool `json:"installable"`
	// Missing 是本机缺少的前置依赖。非空时界面显示原因而不是安装按钮——
	// 让用户点一个注定失败的按钮不如直接告诉他缺什么。
	Missing []ViPrereq `json:"missing,omitempty"`
	Note    string     `json:"note,omitempty"`
}

type ViPrereq struct {
	Binary string `json:"binary"`
	Hint   string `json:"hint"`
}

type ViSessionCreate struct {
	Cwd     string `json:"cwd" validate:"required"`
	Title   string `json:"title"`
	Harness string `json:"harness"`
}

type ViSessionID struct {
	ID string `json:"id" validate:"required"`
}

type ViSessionRename struct {
	ID    string `json:"id" validate:"required"`
	Title string `json:"title" validate:"required"`
}

type ViPoolInfo struct {
	Size   int      `json:"size"`
	Active int      `json:"active"`
	Order  []string `json:"order"`
}

type ViPoolSize struct {
	Size int `json:"size" validate:"required,min=1,max=16"`
}

type ViAccountCapture struct {
	Harness string `json:"harness" validate:"required"`
	Label   string `json:"label"`
}

type ViAccountRef struct {
	Harness string `json:"harness" validate:"required"`
	ID      string `json:"id" validate:"required"`
}

type ViProxy struct {
	URL string `json:"url"`
}

type ViProviderReq struct {
	Harness string `json:"harness" validate:"required"`
	BaseURL string `json:"baseUrl"`
	APIKey  string `json:"apiKey"`
}

type ViHarnessRef struct {
	Harness string `json:"harness" validate:"required"`
}
